e2e/stack.go

```diff
@@ -0,0 +1,213 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	tmdb "github.com/cometbft/cometbft-db"
+	opgenesis "github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer"
+	"github.com/polymerdao/monomer/genesis"
+	"github.com/polymerdao/monomer/node"
+	"github.com/polymerdao/monomer/testutil/testapp"
+	"github.com/polymerdao/monomer/utils"
+	rolluptypes "github.com/polymerdao/monomer/x/rollup/types"
+	"github.com/sourcegraph/conc"
+)
+
+type EventListener interface {
+	OPEventListener
+
+	OnCmdStart(programName string, stdout, stderr io.Reader)
+	OnCmdStopped(programName string, err error)
+}
+
+type Stack struct {
+	l1Address        string
+	monomerAddress   string
+	opNodeAddress    string
+	contractsRootDir string
+	eventListener    EventListener
+	l1BlockTime      time.Duration
+}
+
+// New assumes all ports are available and that all paths exist and are valid.
+func New(l1Address, monomerAddress, opNodeAddress, contractsRootDir string, l1BlockTime time.Duration, eventListener EventListener) *Stack {
+	return &Stack{
+		l1Address:        l1Address,
+		monomerAddress:   monomerAddress,
+		opNodeAddress:    opNodeAddress,
+		contractsRootDir: contractsRootDir,
+		eventListener:    eventListener,
+		l1BlockTime:      l1BlockTime,
+	}
+}
+
+func (s *Stack) Run(parentCtx context.Context) (err error) {
+	var wg conc.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancelCause(parentCtx)
+	defer func() {
+		cancel(err)
+		err = utils.Cause(ctx)
+	}()
+
+	const numConnectAttempts = 20
+
+	// Run anvil.
+	_, l1Port, err := net.SplitHostPort(s.l1Address)
+	if err != nil {
+		return fmt.Errorf("split l1 address into host and port: %v", s.l1Address)
+	}
+	wg.Go(func() {
+		cancel(fmt.Errorf("run cmd: %v", s.runCmd(
+			ctx,
+			"anvil",
+			"--port", l1Port,
+			"--order", "fifo",
+			"--disable-block-gas-limit",
+			"--gas-price", "0",
+			"--block-time", fmt.Sprint(s.l1BlockTime.Seconds()),
+		)))
+	})
```

Comment:
The nesting function calls & implicit hanging function made this a bit hard to parse.

Intent is probably clearer with something like this: 

```suggestion
    wg.Go(func() {
		anvilErr := s.runCmd(
			ctx,
			"anvil",
			"--port", l1Port,
			"--order", "fifo",
			"--disable-block-gas-limit",
			"--gas-price", "0",
			"--block-time", fmt.Sprint(s.l1BlockTime.Seconds()),
		)
		if anvilErr != nil {
			cancel(fmt.Errorf("run cmd: %v", anvilErr))
		}
	})
```

Reply:
Nice catch! There's also a bug in my original version where we'll end up wrapping a `nil` error. Should be fixed in the latest commit. That proves it was indeed too confusing.